File: get-cookies-script/src/kimi.ts
import CDP from "chrome-remote-interface";
import { launchChrome, saveCookies } from "./chrome";

const TARGET_URL = "https://www.kimi.com/code/console";
const CDP_PORT = 9223;
const PROFILE_NAME = "kimi";

async function validateToken(token: string): Promise<boolean> {
  try {
    const response = await fetch(
      "https://www.kimi.com/apiv2/kimi.gateway.billing.v1.BillingService/GetUsages",
      {
        method: "POST",
        headers: {
          accept: "*/*",
          authorization: `Bearer ${token}`,
          "content-type": "application/json",
          "connect-protocol-version": "1",
          origin: "https://www.kimi.com",
          referer: "https://www.kimi.com/code/console",
          "x-language": "zh-CN",
          "x-msh-platform": "web",
          "x-msh-version": "1.0.0",
          "r-timezone": "Asia/Shanghai",
        },
        body: JSON.stringify({ scope: ["FEATURE_CODING"] }),
      }
    );

    const data = await response.json();

    if (data.code === "unauthenticated") {
      return false;
    }

    return !!data.usages;
  } catch {
    return false;
  }
}

export async function fetchKimiCookies(): Promise<void> {
  const outputDir = process.env.COOKIES_OUTPUT_PATH || process.cwd();
  let client: Awaited<ReturnType<typeof CDP>> | null = null;
  let cleanupChrome: (() => void) | null = null;

  try {
    console.log(`📂 Cookies 输出目录: ${outputDir}`);
    cleanupChrome = await launchChrome({ port: CDP_PORT, profileName: PROFILE_NAME });

    console.log("🔌 连接到 Chrome DevTools Protocol...");
    client = await CDP({ port: CDP_PORT });

    const { Network, Page } = client;

    await Network.enable();
    await Page.enable();

    console.log("📄 打开 Kimi coding console 页面...");
    await Page.navigate({ url: TARGET_URL });

    console.log("⏳ 等待页面加载...");
    await Page.loadEventFired();

    console.log("\n" + "=".repeat(60));
    console.log("✅ 浏览器已打开，请在浏览器中完成登录");
    console.log("🔍 每 3 秒检测一次登录状态...");
    console.log("=".repeat(60) + "\n");

    let isLoggedIn = false;
    let attempts = 0;
    const maxAttempts = 30;

    while (!isLoggedIn && attempts < maxAttempts) {
      attempts++;

      try {
        const cookies = await Network.getCookies({ urls: ["https://www.kimi.com/"] });
        const kimiAuthCookie = cookies.cookies.find((c) => c.name === "kimi-auth");

        const kimiCookies = cookies.cookies.filter(
          (c) => c.domain.includes("kimi") || c.domain.includes("moonshot")
        );

        if (kimiCookies.length > 0) {
          console.log(`[${attempts}] 找到 ${kimiCookies.length} 个 kimi 相关 cookies:`);
          kimiCookies.forEach((c) => {
            console.log(`    - ${c.name} (httpOnly: ${c.httpOnly}, secure: ${c.secure})`);
          });
        }

        if (kimiAuthCookie) {
          const token = kimiAuthCookie.value;
          const tokenPreview = token.substring(0, 20) + "...";

          console.log(`[${attempts}] ✅ 检测到 kimi-auth cookie: ${tokenPreview}`);
          console.log(`[${attempts}] 调用 usage 接口验证...`);

          const isValid = await validateToken(token);

          if (isValid) {
            console.log(`[${attempts}] ✅ usage 接口调用成功！登录成功！`);
            isLoggedIn = true;
            break;
          } else {
            console.log(`[${attempts}] ⚠️  usage 接口返回失败，token 无效，继续等待...`);
          }
        } else {
          console.log(`[${attempts}] ⚠️  未检测到 kimi-auth cookie，继续等待...`);
        }
      } catch (error) {
        console.warn(`[${attempts}] 检查时出错:`, error);
      }

      await new Promise((resolve) => setTimeout(resolve, 3000));
    }

    if (!isLoggedIn) {
      console.log("\n❌ 登录检测超时，请重试");
      process.exit(1);
    }

    console.log("\n✅ 检测到登录成功！");
    console.log("🍪 正在获取 Cookies...\n");

    const cookiesResult = await Network.getCookies({ urls: ["https://www.kimi.com/"] });
    const relevantCookies = cookiesResult.cookies.filter(
      (cookie) => cookie.domain.includes("kimi") || cookie.domain.includes("moonshot")
    );

    await saveCookies(relevantCookies, outputDir);
  } catch (error) {
    console.error("❌ 发生错误:", error);
    process.exit(1);
  } finally {
    console.log("🏁 登录脚本结束，开始资源清理");
    if (client) {
      await client.close();
    }
    if (cleanupChrome) {
      console.log("🧹 关闭 Chrome 浏览器...");
      cleanupChrome();
    }
  }
}

File: src-tauri/src/providers/kimi.rs
use std::fs;
use std::path::{Path, PathBuf};
use std::pin::Pin;
use std::future::Future;

use serde::{Deserialize, Serialize};
use tauri::{
    menu::{CheckMenuItem, IsMenuItem, MenuItem, PredefinedMenuItem},
    AppHandle, Wry,
};

use crate::error::{AppError, Result};
use super::{format_progress_bar, format_iso_time, Provider, UsageInfo};

pub const KIMI_ID: &str = "kimi-coding-plan";
pub const KIMI_DISPLAY_NAME: &str = "Kimi";
pub const KIMI_LOGIN_ARG: &str = "kimi";
pub const KIMI_AUTH_TOKEN: &str = "kimi-auth";

pub struct KimiProvider;

impl Provider for KimiProvider {
    fn id(&self) -> &'static str { KIMI_ID }
    fn display_name(&self) -> &'static str { KIMI_DISPLAY_NAME }
    fn login_script_arg(&self) -> &'static str { KIMI_LOGIN_ARG }
    fn auth_token_name(&self) -> &'static str { KIMI_AUTH_TOKEN }
    
    fn fetch_usage(&self, cookie_path: PathBuf) -> Pin<Box<dyn Future<Output = Result<UsageInfo>> + Send + 'static>> {
        Box::pin(async move {
            let cookies = read_cookies(&cookie_path)?;
            let auth_token = find_cookie_value(&cookies, KIMI_AUTH_TOKEN)
                .ok_or_else(|| AppError::Auth("未找到 kimi-auth cookie，请重新登录".to_string()))?
                .to_string();
            
            let client = reqwest::Client::builder()
                .build()
                .map_err(|e| AppError::Http(e.to_string()))?;
            
            let response = client
                .post("https://www.kimi.com/apiv2/kimi.gateway.billing.v1.BillingService/GetUsages")
                .header("accept", "*/*")
                .header("authorization", format!("Bearer {}", auth_token))
                .header("content-type", "application/json")
                .header("connect-protocol-version", "1")
                .header("origin", "https://www.kimi.com")
                .header("referer", "https://www.kimi.com/code/console")
                .header("x-language", "zh-CN")
                .header("x-msh-platform", "web")
                .header("x-msh-version", "1.0.0")
                .header("r-timezone", "Asia/Shanghai")
                .header("User-Agent", "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36")
                .json(&serde_json::json!({"scope": ["FEATURE_CODING"]}))
                .send()
                .await
                .map_err(|e| AppError::Http(e.to_string()))?;
            
            let response_text = response
                .text()
                .await
                .map_err(|e| AppError::Http(e.to_string()))?;
            
            if let Ok(error_response) = serde_json::from_str::<serde_json::Value>(&response_text) {
                if error_response.get("code").and_then(|c| c.as_str()) == Some("unauthenticated") {
                    let _ = fs::remove_file(cookie_path);
                    return Err(AppError::Auth("Token 已失效，请重新登录".to_string()));
                }
            }
            
            let api_response: KimiApiResponse = serde_json::from_str(&response_text)
                .map_err(|e| AppError::Parse(format!("{} - {}", e, response_text)))?;
            
            let usage = api_response
                .usages
                .iter()
                .find(|u| u.scope == "FEATURE_CODING")
                .ok_or(AppError::Parse("未找到 CODING 使用情况".to_string()))?;
            
            let info = build_usage_info(usage)?;
            Ok(UsageInfo::Kimi(info))
        })
    }
    
    fn render_menu_items(
        &self,
        app: &AppHandle,
        usage: &UsageInfo,
        is_selected: bool,
    ) -> Vec<Box<dyn IsMenuItem<Wry>>> {
        let UsageInfo::Kimi(info) = usage else { return vec![] };
        let mut items: Vec<Box<dyn IsMenuItem<Wry>>> = Vec::new();
        
        items.push(Box::new(CheckMenuItem::with_id(
            app,
            format!("select-{}", self.id()),
            format!("{} Coding Plan", self.display_name()),
            true,
            is_selected,
            None::<&str>,
        ).unwrap()));
        
        items.push(Box::new(MenuItem::with_id(
            app,
            format!("{}-hourly-title", self.id()),
            format!("Token 额度（每 {} 小时）", info.hourly_window),
            false,
            None::<&str>,
        ).unwrap()));
        items.push(Box::new(MenuItem::with_id(
            app,
            format!("{}-hourly-bar", self.id()),
            format_progress_bar(info.hourly_percentage),
            false,
            None::<&str>,
        ).unwrap()));
        items.push(Box::new(MenuItem::with_id(
            app,
            format!("{}-hourly-reset", self.id()),
            format!("重置: {}", info.hourly_reset_time),
            false,
            None::<&str>,
        ).unwrap()));
        
        items.push(Box::new(MenuItem::with_id(
            app,
            format!("{}-sep", self.id()),
            "-".repeat(25),
            false,
            None::<&str>,
        ).unwrap()));
        
        items.push(Box::new(MenuItem::with_id(
            app,
            format!("{}-weekly-title", self.id()),
            "Token 额度（每周）",
            false,
            None::<&str>,
        ).unwrap()));
        items.push(Box::new(MenuItem::with_id(
            app,
            format!("{}-weekly-bar", self.id()),
            format_progress_bar(info.weekly_percentage),
            false,
            None::<&str>,
        ).unwrap()));
        items.push(Box::new(MenuItem::with_id(
            app,
            format!("{}-weekly-reset", self.id()),
            format!("重置: {}", info.weekly_reset_time),
            false,
            None::<&str>,
        ).unwrap()));
        items.push(Box::new(PredefinedMenuItem::separator(app).unwrap()));
        
        items
    }
}

pub static KIMI: KimiProvider = KimiProvider;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct KimiUsageInfo {
    pub hourly_total: i64,
    pub hourly_used: i64,
    pub hourly_remaining: i64,
    pub hourly_percentage: f64,
    pub hourly_reset_time: String,
    pub hourly_window: i64,
    pub weekly_total: i64,
    pub weekly_used: i64,
    pub weekly_remaining: i64,
    pub weekly_percentage: f64,
    pub weekly_reset_time: String,
}

#[derive(Debug, Serialize, Deserialize)]
struct KimiApiResponse {
    usages: Vec<KimiUsage>,
}

#[derive(Debug, Serialize, Deserialize)]
struct KimiUsage {
    scope: String,
    detail: KimiUsageDetail,
    limits: Vec<KimiUsageLimit>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
struct KimiUsageDetail {
    #[serde(rename = "limit")]
    limit_value: String,
    used: String,
    remaining: String,
    #[serde(rename = "resetTime")]
    reset_time: String,
}

#[derive(Debug, Serialize, Deserialize)]
struct KimiUsageLimit {
    window: KimiWindow,
    detail: KimiUsageLimitDetail,
}

#[derive(Debug, Serialize, Deserialize)]
struct KimiWindow {
    duration: i64,
    #[serde(rename = "timeUnit")]
    time_unit: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
struct KimiUsageLimitDetail {
    #[serde(rename = "limit")]
    limit_value: String,
    remaining: String,
    #[serde(rename = "resetTime")]
    reset_time: String,
}

fn build_usage_info(usage: &KimiUsage) -> Result<KimiUsageInfo> {
    let limit_detail = usage
        .limits
        .first()
        .ok_or(AppError::Parse("未找到 limits 数据".to_string()))?
        .detail
        .clone();
    let window = &usage.limits.first().unwrap().window;
    
    let hourly_total = limit_detail
        .limit_value
        .parse::<i64>()
        .map_err(|e| AppError::Parse(format!("解析 limit 失败: {}", e)))?;
    let hourly_remaining = limit_detail
        .remaining
        .parse::<i64>()
        .map_err(|e| AppError::Parse(format!("解析 remaining 失败: {}", e)))?;
    let hourly_used = hourly_total - hourly_remaining;
    let hourly_percentage = if hourly_total > 0 {
        (hourly_used as f64 / hourly_total as f64) * 100.0
    } else {
        0.0
    };
    let hourly_reset_time = format_iso_time(&limit_detail.reset_time);
    let hourly_window = window.duration / 60;
    
    let weekly_detail = usage.detail.clone();
    let weekly_total = weekly_detail
        .limit_value
        .parse::<i64>()
        .map_err(|e| AppError::Parse(format!("解析 weekly limit 失败: {}", e)))?;
    let weekly_used = weekly_detail
        .used
        .parse::<i64>()
        .map_err(|e| AppError::Parse(format!("解析 weekly used 失败: {}", e)))?;
    let weekly_remaining = weekly_detail
        .remaining
        .parse::<i64>()
        .map_err(|e| AppError::Parse(format!("解析 weekly remaining 失败: {}", e)))?;
    let weekly_percentage = if weekly_total > 0 {
        (weekly_used as f64 / weekly_total as f64) * 100.0
    } else {
        0.0
    };
    let weekly_reset_time = format_iso_time(&weekly_detail.reset_time);
    
    Ok(KimiUsageInfo {
        hourly_total,
        hourly_used,
        hourly_remaining,
        hourly_percentage,
        hourly_reset_time,
        hourly_window,
        weekly_total,
        weekly_used,
        weekly_remaining,
        weekly_percentage,
        weekly_reset_time,
    })
}

fn read_cookies(cookie_path: &Path) -> Result<Vec<serde_json::Value>> {
    if !cookie_path.exists() {
        return Err(AppError::Auth("未找到 cookies，请先登录".to_string()));
    }
    let cookie_content = std::fs::read_to_string(cookie_path)?;
    let cookies: Vec<serde_json::Value> = serde_json::from_str(&cookie_content)?;
    Ok(cookies)
}

fn find_cookie_value<'a>(cookies: &'a [serde_json::Value], name: &str) -> Option<&'a str> {
    cookies
        .iter()
        .find(|c| c["name"] == name)
        .and_then(|c| c["value"].as_str())
}
